frontend/public/components/utils/firehose.jsx

```diff
@@ -208,6 +223,7 @@ Firehose.propTypes = {
     selector: PropTypes.object,
     fieldSelector: PropTypes.string,
     isList: PropTypes.bool,
-    optional: PropTypes.bool,
+    optional: PropTypes.bool, // do not block children-rendering while resource is still being loaded
+    ignoreIfMissing: PropTypes.bool, // do not fail if the resource is missing (HTTP 404)
```

Comment:
Someone can opinionate that `optional` can be reused for that feature, as the difference is quite small. 
I prefer to keep semantics of `optional` as it is, so existing functionality will not be broken.

If not clear otherwise: if `optional=true`, the children are rendered no matter if loading is still in progress. And eventually causes error 404 afterwards.

With `ignoreIfMissing=true` only, the rendering is postponed till data are loaded or errored. But 404 is not a blocker for rendering. A child component can still check the error code, but the overall error checking is not affected.

It makes sense to set these props individually same as both together.

Reply:
Having two similarly-named properties that behave almost the same, but subtly different, makes for a confusing API. I'm still unclear on the difference even after the explaination :) Can't callers check the `loaded` value if they don't want to render when loading?

I'd prefer either:

1. Using `optional` as-is
2. Changing `optional` if we think there is better behavior (as long as we can do so without breaking current use)

Reply:
I only see 5 places where `optional` is used today, so changing behavior is an option as long we can test / update those places.

Again, I'm still unclear on what's different. Can you give a specific example?

Reply:
I found that when reusing `StatusBox` component.

If `optional=true` for a single (non-list) object and 404 occurs, the `Firehose` injects top-level `loadError` with 404.

With the change proposed in this PR, the `ignoreIfMissing` resource will not affect the top-level `loadError` with 404.

My component structure is like:
```
<Firehose>
  <StatusBox>
     <MyComponent>
```

I can change behaviour in `StatusBox` by ignoring the top-level loadError and checking loadError on every loaded resource, but this is more complicated and would introduce more confusion.
Or I can re-implement StatusBox, but this goes against reusability.

Again, expected behaviour by this PR is:
- `optional=true` ==> do not postpone rendering but fail on 404
- `ignoreIfMissing=true` ==> postpone rendering but ignore 404
- both set to true ==> do not block rendering and do not fail if 404 - the resource is really optional we do not care about it's availability.

If we change semantics of `optional` to act like the third option above only, it will be difficult to implement scenario for immediate (non-blocking) rendering of a page/detail and failing asynchronously if missing.
My PRs do not need this feature but I am not sure whether this is not already expected on other places.

Reply:
> If `optional=true` for a single (non-list) object and 404 occurs, the `Firehose` injects top-level `loadError` with 404.

That sounds like a bug. Are we sure that's how `optional` is intended to behave? Can we simply fix `optional`? :)

Reply:
Such a "fix" would be simple.

I don't see `optional` defined anywhere except just the code.
It was my natural understanding that `optional` should ignore 404. But I am not sure whether something is not already relying on recent semantics.

Reply:
I don't know the history, but I believe it was originally added to avoid 403 errors when you don't have permission. I don't see a reason to handle 404 the same way, though.

The places to check are:

* frontend/public/components/resource-quota.jsx
* frontend/public/components/operator-lifecycle-manager/k8s-resource.tsx
* frontend/public/components/RBAC/role.jsx
* frontend/public/components/RBAC/bindings.jsx
* frontend/public/components/cluster-settings/cluster-settings.tsx

I don't see any other usages.